### DISCUSSION TITLE
How do I setup a "catchall" for a captive-portal access point?

### DISCUSSION THREAD

**Discussion by Nikorasu** [miguelgrinberg/microdot#141] [Q&A] [answered]
I'm trying to setup a Pico W as a captive-portal access point, so that when users connect, they're automatically redirected to the page it's hosting. I actually managed to do this before using Phew, but since that was rather buggy, I've recently converted to Microdot.  I figured I'd just use the same DNS catchall server from Phew, tho I'm not sure if it's compatible with Microdot, since there's no catchall() function?

This code is usually what's required to enable a captive-portal (already converted most of it to microdot):
```py
# microsoft windows redirects
@server.get("/ncsi.txt")
def hotspot(request):
    print("AP ncsi.txt request received")
    return "", 200

@server.get("/connecttest.txt")
def hotspot(request):
    print("AP connecttest.txt request received")
    return "", 200

@server.get("/redirect")
def hotspot(request):
    print("AP redirect request received")
    return redirect(f"http://{DOMAIN}/", 302)

# android redirects
@server.get("/generate_204")
def hotspot(request):
    print("AP generate_204 request received")
    return redirect(f"http://{DOMAIN}/", 302)

# apple redir
@server.get("/hotspot-detect.html")
def hotspot(request):
    print("AP hotspot-detect.html request received")
    return render_template('guestbook.html')

@server.catchall()  # This is how other webservers do it, is there an equivalent way with Microdot?
def catch_all(request):
    print("***CATCHALL***\n" + str(request))
    return redirect("http://" + DOMAIN + "/")
```
At the moment, I'm able to connect to my Pico W, but I have to manually enter 192.168.4.1 to go to it's page, and I should be able to enter anything into the URL and be redirected to the hosted page, or click a notification and be taken to it.

**Comment by miguelgrinberg** [comment]
Have you tried:

```python
@server.get('/<path:path>')
def catch_all(request, path):
    print("***CATCHALL***\n" + str(request))
    return redirect("http://" + DOMAIN + "/")
```

**Reply by Nikorasu** [reply]
Actually yes, I tried that, but it still won't redirect to my hosted page, when the user tries to access the domain or other addresses on the AP, nor do the captive-portal notifications pop up. I can only access my page by going to 192.168.4.1 manually, which is the address the device gave itself for AP mode, I guess.

I assume I also need a dns-catchall server running on port 53, so I've been using the simple one that normally comes with Phew:
```py
import uasyncio, usocket

async def _handler(socket, ip_address):
  while True:
    try:
      yield uasyncio.core._io_queue.queue_read(socket)
      request, client = socket.recvfrom(256)
      response = request[:2] # request id
      response += b"\x81\x80" # response flags
      response += request[4:6] + request[4:6] # qd/an count
      response += b"\x00\x00\x00\x00" # ns/ar count
      response += request[12:] # origional request body
      response += b"\xC0\x0C" # pointer to domain name at byte 12
      response += b"\x00\x01\x00\x01" # type and class (A record / IN class)
      response += b"\x00\x00\x00\x3C" # time to live 60 seconds
      response += b"\x00\x04" # response length (4 bytes = 1 ipv4 address)
      response += bytes(map(int, ip_address.split("."))) # ip address parts
      socket.sendto(response, client)
    except Exception as e:
      print(e)

def run_catchall(ip_address, port=53):
  print("> starting catch all dns server on port {}".format(port))

  _socket = usocket.socket(usocket.AF_INET, usocket.SOCK_DGRAM)
  _socket.setblocking(False)
  _socket.setsockopt(usocket.SOL_SOCKET, usocket.SO_REUSEADDR, 1)
  _socket.bind(usocket.getaddrinfo(ip_address, port, 0, usocket.SOCK_DGRAM)[0][-1])

  loop = uasyncio.get_event_loop()
  loop.create_task(_handler(_socket, ip_address))
  ```
I've seen a few alternatives to that, but I'm not sure what I need exactly.
And for more context, here's how I'm running things:
```py
server = Microdot()
ap = network.WLAN(network.AP_IF)
ap.config(essid=SSID, security=0, pm=0xa11140)
ap.active(True)
ip = ap.ifconfig()[0]
dns.run_catchall(ip)
server.run(port=80,debug=True)
```

**Comment by Nikorasu** [comment, marked as answer]
Okay, sorry.. Now I feel a little silly, but I found a solution to my problem.
Turns out I needed to use the Asyncio version of microdot. Which earlier I had figured I might need to use, but then I totally forgot about it until now. ^^;
So ya, essentially the catchall function you suggested does seem to work fine. Thanks!
And actually it turns out, I don't even need the catchall! Just an errorhandler(404) redirecting back to the domain.

Anywho, I guess maybe I should just close this now that I got it working?

**Reply by zdrawkocolic** [reply]
Hi @Nikorasu,
Could you please share your code? Your realization. I'm facing with same problem and I'm pretty new in this. I manage to do with Phew, but I want to realize it with Microdot.
Thank you in advance.
